### PR TITLE
const_eval: update for const_mut_refs and const_refs_to_cell stabilization

### PR DESCRIPTION
**Pull request by RalfJung**
This is the reference PR accompanying https://github.com/rust-lang/rust/pull/129195.

Fixes https://github.com/rust-lang/rust/issues/57349
Fixes https://github.com/rust-lang/rust/issues/80384

### CODE REVIEW

File: src/const_eval.md [hunk 1]
```diff
@@ -38,8 +38,11 @@ to be run.
 * [Closure expressions] which don't capture variables from the environment.
 * Built-in [negation], [arithmetic], [logical], [comparison] or [lazy boolean]
   operators used on integer and floating point types, `bool`, and `char`.
-* Shared [borrow]s, except if applied to a type with [interior mutability].
-* The [dereference operator] except for raw pointers.
+* All forms of [borrow]s, including raw borrows, with one limitation:
+  mutable borrows and shared borrows to values with interior mutability
+  are only allowed to refer to *transient* places. A place is *transient*
+  if it will be deallocated before the end of evaluating the current constant item.
```
Comment (traviscross) [inline src/const_eval.md:44]: How do we think about what a "constant item" means with respect to `const { .. }` blocks?  Would it be better here to say "end of evaluating the current [constant context](https://doc.rust-lang.org/nightly/reference/const_eval.html#const-context)"?

Comment (RalfJung) [reply src/const_eval.md:44]: If "evaluating a const context" is meaningful, then yeah that would work. It sounds a bit odd to me, how can you evaluate a "context"? The context is what the evaluation occurs in.

No strong preference either way.

Comment (traviscross) [reply src/const_eval.md:44]: Maybe "evaluating the contents of the current const context"?  Open to other ideas.  Saying either "evaluating a context" or "evaluating an item" sound about equally odd to me.

The main thrust is that [const context](https://doc.rust-lang.org/nightly/reference/const_eval.html#const-context) is something we define and seems to be closer to what we mean here.

Comment (RalfJung) [reply src/const_eval.md:44]: "evaluating the item" is how we call it in the compiler so I got used to that one. ;)

**Review by traviscross — APPROVED** [review]

**Review by ehuss — APPROVED** [review]
Thanks!

**Pull request merged**